verification/io/io-spec.gobra

```diff
@@ -63,25 +63,42 @@ pure func dp3s_iospec_bio3s_create_guard_traversedseg(currseg IO_seg3) IO_seg3 {
 			History: currseg.History}
 }
 
+ghost
+requires len(currseg.Future) > 0
+decreases
+pure func (dp DataPlaneSpec) dp3s_iospec_bio3s_create_guard_traversedseg_inc(currseg IO_seg3) IO_seg3 {
+	return let uinfo := !currseg.ConsDir ?
+		dp.upd_uinfo(currseg.UInfo, currseg.Future[0]) :
+		currseg.UInfo in
+		IO_seg3_ {
+			AInfo: currseg.AInfo,
+			UInfo: uinfo,
+			ConsDir: currseg.ConsDir,
+			Peer: currseg.Peer,
+			Past: seq[IO_HF]{currseg.Future[0]} ++ currseg.Past,
+			Future: currseg.Future[1:],
+			History: seq[IO_ahi]{currseg.Future[0].Toab()} ++ currseg.History}
+}
```

Comment:
Why are we introducing `dp3s_iospec_bio3s_create_guard_traversedseg_inc`?

Reply:
I would suggest excluding this from this PR and introducing it in the next one (#231). 
Currently, the concretization of the `enter_guard` , `xover_up2down_guard` and `xover_core_guard` is too restrictive. It does not allow for the increment of the hopfield sequence (`future, past, history`) which is needed for externally forwarded packets. This makes `xover_up2down_guard` and `xover_core_guard` unsatisfiable and the external forwarding in the `enter_guard` as well. This can be seen by looking at the function [`inc_seg2()`](https://github.com/viperproject/VerifiedSCION/pull/230/files#diff-b475d495934e1cd3662cc65c72154eca2efb9b33fbfe4f2eb9c06280f27f056cR26). It requires that the `future, past, history` field of the `traversedseg` are incremented. However, `traversedseg` is defined with equality in those fields. To address this, we need  `dp3s_iospec_bio3s_create_guard_traversedseg_inc` for the xover guards and introduce a case distinction (external or internal packet) in the enter guard (this distinction has already been implemented in #231).

Reply:
Aren't you effectively rewriting the IO spec if you do that? If so, we need to find some other solution to guarantee that we preserve all the properties from the high-level model down to the implementation

Reply:
I don't think so. `traversedseg` is a quantified existentially in the IO-spec. `dp3s_iospec_bio3s_create_guard_traversedseg()` is just a helper function to concretize `traversedseg` . This function doesn't exist in IO-spec. Originally, we though `traversedseg` has to be concretized exactly the same way for all guards, so the common term was factored out into a helper function. However, it turns out that we need different `traversedseg` for different guards.

Reply:
I see. I agree with merging this in the next PR. There, we should also document this clearly, to make sure that we will remember this design decision in the future